table/table_test.cc

```diff
@@ -5367,6 +5369,181 @@ TEST_P(
+                        BlockBasedTable::Rep::FilterType::kFullFilter,
+                        BloomFilterPolicy::Mode::kLegacyBloom),
+        std::make_tuple(true, test::kLatestFormatVersion,
+                        BlockBasedTable::Rep::FilterType::kPartitionedFilter,
+                        BloomFilterPolicy::Mode::kFastLocalBloom),
+        std::make_tuple(true, test::kLatestFormatVersion,
+                        BlockBasedTable::Rep::FilterType::kPartitionedFilter,
+                        BloomFilterPolicy::Mode::kStandard128Ribbon),
+        std::make_tuple(true, test::kLatestFormatVersion,
+                        BlockBasedTable::Rep::FilterType::kPartitionedFilter,
+                        BloomFilterPolicy::Mode::kLegacyBloom),
+        std::make_tuple(true, test::kLatestFormatVersion,
+                        BlockBasedTable::Rep::FilterType::kBlockFilter,
+                        BloomFilterPolicy::Mode::kDeprecatedBlock)));
+
+TEST_P(BlockBasedTableReserveFilterConstructionMemoryTest,
+       OptionalFilterConstructionMemoryReservation) {
+  constexpr std::size_t kSizeDummyEntry = 256 * 1024;
+  constexpr std::size_t kCacheCapacity = 8 * 1024 * 1024;
+
+  BlockBasedTableOptions table_options = GetBlockBasedTableOptions();
+  LRUCacheOptions lo;
+  lo.capacity = kCacheCapacity;
+  lo.num_shard_bits = 0;  // 2^0 shard
+  lo.strict_capacity_limit = true;
+  std::shared_ptr<Cache> cache(NewLRUCache(lo));
+  table_options.block_cache = cache;
+  Options options;
+  options.compression = kSnappyCompression;
+  options.table_factory.reset(NewBlockBasedTableFactory(table_options));
+
+  test::StringSink* sink = new test::StringSink();
+  std::unique_ptr<FSWritableFile> holder(sink);
+  std::unique_ptr<WritableFileWriter> file_writer(new WritableFileWriter(
+      std::move(holder), "test_reserve_filter_construction_memory",
+      FileOptions()));
+  ImmutableOptions ioptions(options);
+  MutableCFOptions moptions(options);
+  InternalKeyComparator ikc(options.comparator);
+  IntTblPropCollectorFactories int_tbl_prop_collector_factories;
+
+  ASSERT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+  std::unique_ptr<TableBuilder> builder(options.table_factory->NewTableBuilder(
+      TableBuilderOptions(ioptions, moptions, ikc,
+                          &int_tbl_prop_collector_factories, kSnappyCompression,
+                          options.compression_opts, kUnknownColumnFamily,
+                          "test_cf", -1 /* level */),
+      file_writer.get()));
+
+  const std::size_t num_entries = 1000000;
+
+  for (std::size_t i = 0; i < num_entries; ++i) {
+    std::string key = DBTestBase::Key(static_cast<int>(i));
+    std::string value = "val" + std::to_string(static_cast<int>(i));
+    InternalKey ik(key, i /* sequnce number */, kTypeValue);
+    builder->Add(ik.Encode(), value);
+  }
+
+  ASSERT_OK(builder->Finish());
+  EXPECT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
```

Comment:
I don't see this test checking that memory is actually reserved. What is the intent?

I think we should use a sync point callback at the point of expected peak reservation, or a custom Cache wrapper, to ensure the memory reservation reaches something around what is expected. IMHO, detecting the peak is a sufficient regression test. I'm willing to believe the "area under the curve" is accounted over time by inspection.